<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="description" content="This is a snapshot site for StackOverflow">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>StackOverflow Snapshot (old posts, page 1308) | StackOverflow Snapshot</title>
<link href="assets/css/all-nocdn.css" rel="stylesheet" type="text/css">
<meta name="theme-color" content="#5670d4">
<meta name="generator" content="Nikola (getnikola.com)">
<link rel="canonical" href="https://ohstackoverflow.netlify.app/index-1308.html">
<link rel="prev" href="index-1309.html" type="text/html">
<link rel="next" href="index-1307.html" type="text/html">
<!--[if lt IE 9]><script src="assets/js/html5.js"></script><![endif]-->
</head>
<body>
<a href="#content" class="sr-only sr-only-focusable">Skip to main content</a>

<!-- Menubar -->

<nav class="navbar navbar-inverse navbar-static-top"><div class="container">
<!-- This keeps the margins nice -->
        <div class="navbar-header">
            <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#bs-navbar" aria-controls="bs-navbar" aria-expanded="false">
            <span class="sr-only">Toggle navigation</span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand" href="https://ohstackoverflow.netlify.app/">

                <span id="blog-title">StackOverflow Snapshot</span>
            </a>
        </div>
<!-- /.navbar-header -->
        <div class="collapse navbar-collapse" id="bs-navbar" aria-expanded="false">
            <ul class="nav navbar-nav">
<li>
<a href="archive.html">Archive</a>
                </li>
<li>
<a href="categories/">Tags</a>

                
            </li>
</ul>
<ul class="nav navbar-nav navbar-right">
<div style="display:table;min-height:5rem;min-width:27rem;">
					<div class="input-group" style="display: table-cell;vertical-align: middle;">
						<input id="words" type="text" class="form-control" style="max-width:22rem;" onkeydown="if(event.keyCode==13){btn.click()}"><span class="input-group-btn" style="float:left">
							<button id="btn" class="btn btn-default" type="button" data-toggle="modal" data-target="#myModal">
								<span class="glyphicon glyphicon-search">
							</span></button>
						</span>
					</div>
<!-- /input-group -->
				</div>

				
                
                
                
            </ul>
</div>
<!-- /.navbar-collapse -->
    </div>
<!-- /.container -->
</nav><!-- End of Menubar --><!-- 模态框（Modal） --><div class="modal fade" id="myModal" tabindex="-1" role="dialog" aria-labelledby="myModalLabel" aria-hidden="true">
	<div class="modal-dialog">
		<div class="modal-content">
			<div class="modal-header">
				<button type="button" class="close" data-dismiss="modal" aria-hidden="true">×
				</button>
				<h4 class="modal-title" id="myModalLabel">
					查找结果
				</h4>
			</div>
			<div class="modal-body">
				<div id="search-count" style="min-height:4rem;">
				查找中，请稍后...
				</div>
				<div id="search-result">
				</div>

				
			</div>
			<div class="modal-footer">
				<button type="button" class="btn btn-default" data-dismiss="modal">
					关闭
				</button>
			</div>
		</div>
<!-- /.modal-content -->
	</div>
<!-- /.modal-dialog -->
</div>
<!-- /.modal -->

<div class="container" id="content" role="main">
    <div class="body-content">
        <!--Body content-->
        <div class="row">
            
            

    


    
<div class="postindex">
    <article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="posts/is-there-a-difference-between-throw-and-throw-ex/" class="u-url">Is there a difference between "throw" and "throw ex"?</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                <a href="authors/arya/">Arya</a>
            </span></p>
            <p class="dateline">
            <a href="posts/is-there-a-difference-between-throw-and-throw-ex/" rel="bookmark">
            <time class="published dt-published" datetime="2023-02-18T06:16:50+08:00" itemprop="datePublished" title="2023-02-18 06:16">2023-02-18 06:16</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <p>There are some posts that asks what the difference between those two are
already.<br>
(why do I have to even mention this...)</p>
<p>But my question is different in a way that I am calling "throw ex" in another
error <em>god-like</em> handling method.</p>
<div class="code"><pre class="code literal-block"><span class="nv">public</span><span class="w"> </span><span class="nv">class</span><span class="w"> </span><span class="nv">Program</span><span class="w"> </span>{
<span class="w">    </span><span class="nv">public</span><span class="w"> </span><span class="nv">static</span><span class="w"> </span><span class="nv">void</span><span class="w"> </span><span class="nv">Main</span><span class="ss">(</span><span class="nv">string</span>[]<span class="w"> </span><span class="nv">args</span><span class="ss">)</span><span class="w"> </span>{
<span class="w">        </span><span class="nv">try</span><span class="w"> </span>{
<span class="w">            </span><span class="o">//</span><span class="w"> </span><span class="nv">something</span>
<span class="w">        </span>}<span class="w"> </span><span class="nv">catch</span><span class="w"> </span><span class="ss">(</span><span class="nv">Exception</span><span class="w"> </span><span class="nv">ex</span><span class="ss">)</span><span class="w"> </span>{
<span class="w">            </span><span class="nv">HandleException</span><span class="ss">(</span><span class="nv">ex</span><span class="ss">)</span><span class="c1">;</span>
<span class="w">        </span>}
<span class="w">    </span>}

<span class="w">    </span><span class="nv">private</span><span class="w"> </span><span class="nv">static</span><span class="w"> </span><span class="nv">void</span><span class="w"> </span><span class="nv">HandleException</span><span class="ss">(</span><span class="nv">Exception</span><span class="w"> </span><span class="nv">ex</span><span class="ss">)</span><span class="w"> </span>{
<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="ss">(</span><span class="nv">ex</span><span class="w"> </span><span class="nv">is</span><span class="w"> </span><span class="nv">ThreadAbortException</span><span class="ss">)</span><span class="w"> </span>{
<span class="w">            </span><span class="o">//</span><span class="w"> </span><span class="nv">ignore</span><span class="w"> </span><span class="k">then</span>,
<span class="w">            </span><span class="k">return</span><span class="c1">;</span>
<span class="w">        </span>}
<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="ss">(</span><span class="nv">ex</span><span class="w"> </span><span class="nv">is</span><span class="w"> </span><span class="nv">ArgumentOutOfRangeException</span><span class="ss">)</span><span class="w"> </span>{<span class="w"> </span>
<span class="w">            </span><span class="o">//</span><span class="w"> </span><span class="nv">Log</span><span class="w"> </span><span class="k">then</span>,
<span class="w">            </span><span class="nv">throw</span><span class="w"> </span><span class="nv">ex</span><span class="c1">;</span>
<span class="w">        </span>}
<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="ss">(</span><span class="nv">ex</span><span class="w"> </span><span class="nv">is</span><span class="w"> </span><span class="nv">InvalidOperationException</span><span class="ss">)</span><span class="w"> </span>{
<span class="w">            </span><span class="o">//</span><span class="w"> </span><span class="k">Show</span><span class="w"> </span><span class="nv">message</span><span class="w"> </span><span class="k">then</span>,
<span class="w">            </span><span class="nv">throw</span><span class="w"> </span><span class="nv">ex</span><span class="c1">;</span>
<span class="w">        </span>}
<span class="w">        </span><span class="o">//</span><span class="w"> </span><span class="nv">and</span><span class="w"> </span><span class="nv">so</span><span class="w"> </span><span class="nv">on</span>.
<span class="w">    </span>}
}
</pre></div>

<p>If <code>try &amp; catch</code> were used in the <code>Main</code>, then I would use <code>throw;</code> to rethrow
the error. But in the above simplied code, all exceptions go through
<code>HandleException</code></p>
<p>Does <code>throw ex;</code> has the same effect as calling <code>throw</code> when called inside
<code>HandleException</code>?</p>
<p><br><br></p>
<h2>Answer</h2>
<p>Yes, there is a difference.</p>
<ul>
<li>
<p><code>throw ex</code> resets the stack trace (so your errors would appear to originate from <code>HandleException</code>)</p>
</li>
<li>
<p><code>throw</code> doesn't - the original offender would be preserved.</p>
<div class="code"><pre class="code literal-block"> static void Main(string[] args)
</pre></div>

<p>{
     try
     {
         Method2();
     }
     catch (Exception ex)
     {
         Console.Write(ex.StackTrace.ToString());
         Console.ReadKey();
     }
 }</p>
<p>private static void Method2()
 {
     try
     {
         Method1();
     }
     catch (Exception ex)
     {
         //throw ex resets the stack trace Coming from Method 1 and propogates it to the caller(Main)
         throw ex;
     }
 }</p>
<p>private static void Method1()
 {
     try
     {
         throw new Exception("Inside Method1");
     }
     catch (Exception)
     {
         throw;
     }
 }</p>
</li>
</ul>
<p><br></p>
<h3>Suggest</h3>
<p>(I posted earlier, and @Marc Gravell has corrected me)</p>
<p>Here's a demonstration of the difference:</p>
<div class="code"><pre class="code literal-block"><span class="nt">static</span><span class="w"> </span><span class="nt">void</span><span class="w"> </span><span class="nt">Main</span><span class="o">(</span><span class="nt">string</span><span class="cp">[]</span><span class="w"> </span><span class="nt">args</span><span class="o">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="err">try</span><span class="w"> </span><span class="err">{</span>
<span class="w">        </span><span class="err">ThrowException1()</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">19</span>
<span class="w">    </span><span class="p">}</span><span class="w"> </span><span class="nt">catch</span><span class="w"> </span><span class="o">(</span><span class="nt">Exception</span><span class="w"> </span><span class="nt">x</span><span class="o">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="err">Console.WriteLine("Exception</span><span class="w"> </span><span class="err">1:")</span><span class="p">;</span>
<span class="w">        </span><span class="err">Console.WriteLine(x.StackTrace)</span><span class="p">;</span>
<span class="w">    </span><span class="p">}</span>
<span class="w">    </span><span class="nt">try</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="err">ThrowException2()</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">25</span>
<span class="w">    </span><span class="p">}</span><span class="w"> </span><span class="nt">catch</span><span class="w"> </span><span class="o">(</span><span class="nt">Exception</span><span class="w"> </span><span class="nt">x</span><span class="o">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="err">Console.WriteLine("Exception</span><span class="w"> </span><span class="err">2:")</span><span class="p">;</span>
<span class="w">        </span><span class="err">Console.WriteLine(x.StackTrace)</span><span class="p">;</span>
<span class="w">    </span><span class="p">}</span>
<span class="err">}</span>

<span class="nt">private</span><span class="w"> </span><span class="nt">static</span><span class="w"> </span><span class="nt">void</span><span class="w"> </span><span class="nt">ThrowException1</span><span class="o">()</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="err">try</span><span class="w"> </span><span class="err">{</span>
<span class="w">        </span><span class="err">DivByZero()</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">34</span>
<span class="w">    </span><span class="p">}</span><span class="w"> </span><span class="nt">catch</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="err">throw</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">36</span>
<span class="w">    </span><span class="p">}</span>
<span class="err">}</span>
<span class="nt">private</span><span class="w"> </span><span class="nt">static</span><span class="w"> </span><span class="nt">void</span><span class="w"> </span><span class="nt">ThrowException2</span><span class="o">()</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="err">try</span><span class="w"> </span><span class="err">{</span>
<span class="w">        </span><span class="err">DivByZero()</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">41</span>
<span class="w">    </span><span class="p">}</span><span class="w"> </span><span class="nt">catch</span><span class="w"> </span><span class="o">(</span><span class="nt">Exception</span><span class="w"> </span><span class="nt">ex</span><span class="o">)</span><span class="w"> </span><span class="p">{</span>
<span class="w">        </span><span class="err">throw</span><span class="w"> </span><span class="err">ex</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">43</span>
<span class="w">    </span><span class="p">}</span>
<span class="err">}</span>

<span class="nt">private</span><span class="w"> </span><span class="nt">static</span><span class="w"> </span><span class="nt">void</span><span class="w"> </span><span class="nt">DivByZero</span><span class="o">()</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="err">int</span><span class="w"> </span><span class="err">x</span><span class="w"> </span><span class="err">=</span><span class="w"> </span><span class="err">0</span><span class="p">;</span>
<span class="w">    </span><span class="err">int</span><span class="w"> </span><span class="err">y</span><span class="w"> </span><span class="err">=</span><span class="w"> </span><span class="err">1</span><span class="w"> </span><span class="err">/</span><span class="w"> </span><span class="err">x</span><span class="p">;</span><span class="w"> </span><span class="err">//</span><span class="w"> </span><span class="err">line</span><span class="w"> </span><span class="err">49</span>
<span class="p">}</span>
</pre></div>

<p>and here is the output:</p>
<div class="code"><pre class="code literal-block">Exception 1:
   at UnitTester.Program.DivByZero() in &lt;snip&gt;\Dev\UnitTester\Program.cs:line 49
   at UnitTester.Program.ThrowException1() in &lt;snip&gt;\Dev\UnitTester\Program.cs:line 36
   at UnitTester.Program.TestExceptions() in &lt;snip&gt;\Dev\UnitTester\Program.cs:line 19

Exception 2:
   at UnitTester.Program.ThrowException2() in &lt;snip&gt;\Dev\UnitTester\Program.cs:line 43
   at UnitTester.Program.TestExceptions() in &lt;snip&gt;\Dev\UnitTester\Program.cs:line 25
</pre></div>

<p>You can see that in Exception 1, the stack trace goes back to the
<code>DivByZero()</code> method, whereas in Exception 2 it does not.</p>
<p>Take note, though, that the line number shown in <code>ThrowException1()</code> and
<code>ThrowException2()</code> is the line number of the <code>throw</code> statement, <strong>not</strong> the
line number of the call to <code>DivByZero()</code>, which probably makes sense now that
I think about it a bit...</p>
<h4>Output in Release mode</h4>
<p>Exception 1:</p>
<div class="code"><pre class="code literal-block">at ConsoleAppBasics.Program.ThrowException1()
at ConsoleAppBasics.Program.Main(String[] args)
</pre></div>

<p>Exception 2:</p>
<div class="code"><pre class="code literal-block">at ConsoleAppBasics.Program.ThrowException2()
at ConsoleAppBasics.Program.Main(String[] args)
</pre></div>

<p>Is it maintains the original stackTrace in debug mode only?</p>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="posts/how-to-uncheck-a-radio-button/" class="u-url">How to uncheck a radio button?</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                <a href="authors/arya/">Arya</a>
            </span></p>
            <p class="dateline">
            <a href="posts/how-to-uncheck-a-radio-button/" rel="bookmark">
            <time class="published dt-published" datetime="2023-02-18T06:16:23+08:00" itemprop="datePublished" title="2023-02-18 06:16">2023-02-18 06:16</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <p>I have group of radio buttons that I want to uncheck after an AJAX form is
submitted using jQuery. I have the following function:</p>
<div class="code"><pre class="code literal-block">function clearForm(){
  $('#frm input[type="text"]').each(function(){
      $(this).val("");  
  });
  $('#frm input[type="radio":checked]').each(function(){
      $(this).checked = false;  
  });
 }
</pre></div>

<p>With the help of this function, I can clear the values at the text boxes, but
I can't clear the values of the radio buttons.</p>
<p>By the way, I also tried <code>$(this).val("");</code> but that didn't work.</p>
<p><br><br></p>
<h2>Answer</h2>
<p>either (plain js)</p>
<div class="code"><pre class="code literal-block">this.checked = false;
</pre></div>

<p>or (jQuery)</p>
<div class="code"><pre class="code literal-block">$(this).prop('checked', false);
// Note that the pre-jQuery 1.6 idiom was
// $(this).attr('checked', false);
</pre></div>

<p>See <strong>jQuery prop() help page</strong> for an explanation on the difference between
<em>attr()</em> and <em>prop()</em> and why prop() is now preferable.<br>
prop() was introduced with jQuery 1.6 in May 2011.</p>
<p><br></p>
<h3>Suggest</h3>
<p>You wouldn't need the <code>each</code> function</p>
<div class="code"><pre class="code literal-block"><span class="o">$(</span><span class="s2">"input:radio"</span><span class="o">)</span><span class="p">.</span><span class="nc">attr</span><span class="o">(</span><span class="s2">"checked"</span><span class="o">,</span><span class="w"> </span><span class="nt">false</span><span class="o">);</span>
</pre></div>

<p>Or</p>
<div class="code"><pre class="code literal-block"><span class="o">$(</span><span class="s2">"input:radio"</span><span class="o">)</span><span class="p">.</span><span class="nc">removeAttr</span><span class="o">(</span><span class="s2">"checked"</span><span class="o">);</span>
</pre></div>

<p>The same should also apply to your textbox:</p>
<div class="code"><pre class="code literal-block">$('#frm input[type="text"]').val("");
</pre></div>

<p>But you could improve this</p>
<div class="code"><pre class="code literal-block"><span class="o">$(</span><span class="s1">'#frm input:text'</span><span class="o">)</span><span class="p">.</span><span class="nc">val</span><span class="o">(</span><span class="s2">""</span><span class="o">);</span>
</pre></div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="posts/error-link-fatal-error-lnk1123-failure-during-conversion-to-coff-file-invalid-or-corrupt-after-installing-visual-studio-2012-release-preview/" class="u-url">Error 'LINK : fatal error LNK1123: failure during conversion to COFF: file invalid or corrupt' after installing Visual Studio 2012 Release Preview</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                <a href="authors/arya/">Arya</a>
            </span></p>
            <p class="dateline">
            <a href="posts/error-link-fatal-error-lnk1123-failure-during-conversion-to-coff-file-invalid-or-corrupt-after-installing-visual-studio-2012-release-preview/" rel="bookmark">
            <time class="published dt-published" datetime="2023-02-18T06:15:52+08:00" itemprop="datePublished" title="2023-02-18 06:15">2023-02-18 06:15</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <p>I've installed Visual Studio 2012 Release Preview, and it appears to be fine,
but now when I try to use Visual Studio 2010 to compile C++ projects, I get
the following error message:</p>
<blockquote>
<p>LINK : fatal error LNK1123: failure during conversion to COFF: file invalid
or corrupt</p>
</blockquote>
<p>I'm not 100% sure of this, but it seems to be related to projects that have
<code>.rc</code> (resource) files in them.</p>
<p>I've tried repairing Visual Studio 2010 from Add/Remove programs and
rebooting, but this has no effect.</p>
<p>I also get the same error if I use Visual Studio 2012 RC to compile the C++
projects when set to use the Visual Studio 2010 toolset. Upgrading to the
Visual Studio 2011 toolset fixes the problem (but of course I don't want to do
this for production code).</p>
<p>Update: I've <em>uninstalled Visual Studio 2012</em> , rebooted, and the problem
still persists! Help!</p>
<p><br><br></p>
<h2>Answer</h2>
<p>This MSDN thread explains how to fix it.</p>
<p>To summarize:</p>
<ul>
<li>
<p>Either disable incremental linking, by going to </p>
<div class="code"><pre class="code literal-block">Project Properties
</pre></div>

<p>-&gt; Configuration Properties 
       -&gt; Linker (General) 
          -&gt; Enable Incremental Linking -&gt; "No (/INCREMENTAL:NO)"</p>
</li>
<li>
<p>or install VS2010 SP1.</p>
</li>
</ul>
<p>Edits (@CraigRinger): Note that installing VS 2010 SP1 will <em>remove the 64-bit
compilers</em>. You need to install the VS 2010 SP1 compiler pack to get them
back.</p>
<p>This affects <strong>Microsoft Windows SDK 7.1 for Windows 7 and .NET 4.0</strong> as well
as <strong>Visual Studio 2010</strong>.</p>
<p><br></p>
<h3>Suggest</h3>
<p>If disabling incremental linking doesn't work for you, and turning off "Embed
Manifest" doesn't work either, then search your path for multiple versions of
CVTRES.exe.</p>
<p>By debugging with the /VERBOSE linker option I found the linker was writing
that error message when it tried to invoke cvtres and it failed.</p>
<p>It turned out that I had two versions of this utility in my path. One at
<code>C:\Program Files (x86)\Microsoft Visual Studio 10.0\VC\BIN\cvtres.exe</code> and
one at <code>C:\Windows\Microsoft.NET\Framework\v4.0.30319\cvtres.exe</code>. After
VS2012 install, the VS2010 version of cvtres.exe will no longer work. If
that's the first one in your path, and the linker decides it needs to convert
a .res file to COFF object format, the link will fail with LNK1123.</p>
<p>(Really annoying that the error message has nothing to do with the actual
problem, but that's not unusual for a Microsoft product.)</p>
<p>Just delete/rename the older version of the utility, or re-arrange your PATH
variable, so that the version that works comes first.</p>
<p>Be aware that for x64 tooling builds you may also have to check <code>C:\Program
Files (x86)\Microsoft Visual Studio 10.0\VC\bin\amd64</code> where there is another
<code>cvtres.exe</code>.</p>
    </div>
    </article>
</div>

        <nav class="postindexpager"><ul class="pager">
<li class="previous">
                <a href="index-1309.html" rel="prev">Newer posts</a>
            </li>
            <li class="next">
                <a href="index-1307.html" rel="next">Older posts</a>
            </li>
        </ul></nav>
</div>
        <!--End of body content-->

        <footer id="footer">
            Contents © 2023         Go to StackOverflow Chinese Site  <a href="http://stackoverflow.ink">StackOverflow-ZH</a>  
            
        </footer>
</div>
</div>


            <script src="assets/js/all-nocdn.js"></script><!-- fancy dates --><script>
    moment.locale("en");
    fancydates(0, "YYYY-MM-DD HH:mm");
    </script><!-- end fancy dates --><script>
    baguetteBox.run('div#content', {
        ignoreClass: 'islink',
        captions: function(element) {
            return element.getElementsByTagName('img')[0].alt;
    }});
    </script><script src="assets/js/search.js"></script>
</body>
</html>
